### PR TITLE
*: implement priority control for OOM Action (#21170)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21170 to release-4.0

---

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/18322 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch 4.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- No release note

### CODE REVIEW

**Comment by wshwsh12** [comment]
Need wait rateLimitAction cherry-pick to 4.0 branch.

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

File: util/memory/tracker_test.go [hunk 1]
```diff
@@ -341,3 +342,38 @@ func BenchmarkConsume(b *testing.B) {
...
+
+	tracker = NewTracker(1, 1)
+	tracker.actionMu.actionOnExceed = nil
+	n := 100
+	actions := make([]*mockAction, n)
+	for i := 0; i < n; i++ {
+		actions[i] = &mockAction{priority: int64(i)}
+	}
+
+	randomSuffle := make([]int, n)
```
Comment (Yisaer) [inline util/memory/tracker_test.go:359]: typo `shuffle`

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

File: util/memory/tracker_test.go (hunk 1 above)
Comment (wshwsh12) [reply util/memory/tracker_test.go]: addressed.

**Review by Yisaer — COMMENTED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**